text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,745 @@
+```
+
+Alternatively the `cdk.out` directory from a previous synth can be used as well:
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCdkOut("/path/to/cdk.out");
+```
+
+#### Actions - synth
+
+The `synth` action takes a Cloud Executable and synthesizes it in to a `CloudAssembly`.
+
+```ts
+const cx: ICloudExecutable =  await cdk.synth(cx, { stacks: ['MyTestStack'] });
+```
+
+It returns a Cloud Executable that is cached on disk and can be used as input for other actions.
+
+#### Actions - bootstrap
+
+The `bootstrap` action takes in a list of environments and bootstraps each environment with the provided configuration:
+
+```ts
+const envs: string[] = cx.environments();
+await cdk.bootstrap(envs, { trust: ['000000000000', '111111111111'] });
+```
+
+It returns a list of stack deployments.
+
+#### Errors
+
+All errors are thrown as exceptions and the integrator must be prepared to handle them.
+Expected errors are structured data and extend the `IoMessage` interface.
+
+```ts
+interface ToolkitError extends IoMessage {
+  source: 'toolkit' | 'user';
+  type: 'authentication' | 'cloudformation' | 'validation' ...;
+  fqn?: string;
+}
+```
+
+The toolkit might throw other exceptions.
+These are bugs and you should report them by [raising an issue](https://github.com/aws/aws-cdk/issues/new?assignees=&labels=bug%2Cneeds-triage&projects=&template=bug-report.yml&title=%28toolkit%29%3A+Untyped+Exception+%28short+description%29).
+To assist integrators with detecting the type of error, the following helper methods are available.
+Even though errors are typed, you should not rely on `instanceof` checks because it can behave unexpectedly when working with multiple copies of the same package. [TODO: insert reference link]
+
+```ts
+try {
+  await cdk.deploy(...);
+} catch (e: unknown) {
+  if (ToolkitError.isToolkitError(error)) {
+    // handle toolkit errors
+  } else if (ToolkitError.isUserError(e)) {
+    // handle user code errors
+  }
+  
+  if (ToolkitError.isCdkError(e)) {
+    // handle any cdk errors
+  }
+}
+```
+
+##### Recoverable Errors [TODO: Naming. All errors in JS are recoverable. Ruby calls this rescue/retry]
+
+  [Alternative: Do not do this at all]
+  [Alternative: Formalize a retry budget]
+
+While most errors are terminal, it's possible to recover from some.
+For example if an authentication token is expired, an integrator may prompt the user to re-authenticate via a browser login, before continuing with the action.
+With recoverable errors, the integrator can return to normal program flow after an issue has been addressed.
+
+In the [Programmatic Toolkit], recoverable errors are implemented as a special type of request.
+By default, errors are not attempted to recover.
+
+```ts
+interface RecoverableError extends IoRequest<T, U> {
+  level: 'error';
+  defaultResponse: {
+    retry: boolean; // indicates if the block should retried
+    // ... other response values
+  };
+  data: {
+    isRecoverable: true;
```

Comment:
I wouldn't have expected this to be part of `data`, but part of the top-level struct?